zkevm-circuits/src/evm_circuit/util/common_gadget.rs

```diff
@@ -99,3 +100,91 @@ impl<F: FieldExt> SameContextGadget<F> {
         Ok(())
     }
 }
+
+#[derive(Clone, Debug)]
+pub(crate) struct TransferWithGasFeeGadget<F> {
+    sub_sender_balance: AddWordsGadget<F, 3>,
+    add_receiver_balance: AddWordsGadget<F, 2>,
+}
+
+impl<F: FieldExt> TransferWithGasFeeGadget<F> {
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        sender_address: Expression<F>,
+        receiver_address: Expression<F>,
+        value: Word<F>,
+        gas_fee: Word<F>,
+        is_persistent: Expression<F>,
+        rw_counter_end_of_reversion: Expression<F>,
+    ) -> Self {
+        let sender_balance = cb.query_word();
+        let receiver_balance_prev = cb.query_word();
+
+        // Subtract sender balance by value and gas_fee
+        let sub_sender_balance = AddWordsGadget::construct(
+            cb,
+            [sender_balance.clone(), value.clone(), gas_fee],
+        );
+        cb.require_zero(
+            "Sender has sufficient balance",
+            sub_sender_balance.carry().expr(),
+        );
+
+        // Add receiver balance by value
+        let add_receiver_balance = AddWordsGadget::construct(
+            cb,
+            [receiver_balance_prev.clone(), value],
+        );
+        cb.require_zero(
+            "Receiver has too much balance",
+            add_receiver_balance.carry().expr(),
+        );
+
+        let sender_balance_prev = sub_sender_balance.sum();
+        let receiver_balance = add_receiver_balance.sum();
+
+        // Write with possible reversion
+        for (address, balance, balance_prev) in [
+            (sender_address, &sender_balance, sender_balance_prev),
+            (receiver_address, receiver_balance, &receiver_balance_prev),
+        ] {
+            cb.account_write_with_reversion(
+                address,
+                AccountFieldTag::Balance,
+                balance.expr(),
+                balance_prev.expr(),
+                is_persistent.clone(),
+                rw_counter_end_of_reversion.clone(),
+            );
+        }
+
+        Self {
+            sub_sender_balance,
+            add_receiver_balance,
+        }
+    }
+
+    pub(crate) fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
```

Comment:
Why using `Region` and `offset` if you can simplify with the `Layouter`?

Is it easier to manage gadgets with the region + offset approach?

Reply:
Ah I have missed this one. Yes using `Region` and `offset` is easier here because we are sharing a lot of cells, if using layouter we might need to calculate offset back, and leads to more complicated codes when assignment.

Reply:
Hmmm. I see. You don't want to enable equalty constraints for all the things you copy right?

Makes sense then.  :)